### PR TITLE
NAS-115711: Support CPU pinning

### PR DESCRIPTION
**Pull request by heppu**
UI for already merged CPU pinning support in middleware. https://github.com/truenas/middleware/pull/9045

### CODE REVIEW

File: src/app/pages/vm/vm-form/vm-form.component.ts [hunk 1]
```diff
@@ -142,6 +142,19 @@ export class VmFormComponent implements FormConfiguration {
           tooltip: helptext.threads.tooltip,
           validation: [Validators.required, Validators.min(1), this.cpuValidator('threads')],
         },
+        {
+          type: 'input',
+          name: 'cpuset',
+          inputType: 'string',
```
Comment (AlexKarpov98) [inline src/app/pages/vm/vm-form/vm-form.component.ts:148]: Please remove `inputType: 'string'`

File: src/app/pages/vm/vm-form/vm-form.component.ts [hunk 2]
```diff
@@ -169,6 +182,13 @@ export class VmFormComponent implements FormConfiguration {
           blurEvent: () => this.memoryBlur(),
           parent: this,
         },
+        {
+          type: 'input',
+          name: 'nodeset',
+          inputType: 'string',
```
Comment (AlexKarpov98) [inline src/app/pages/vm/vm-form/vm-form.component.ts:188]: Same here remove `inputType: 'string'`

Comment (heppu) [reply src/app/pages/vm/vm-form/vm-form.component.ts]: Done

File: src/app/pages/vm/vm-form/vm-form.component.ts [hunk 3]
```diff
@@ -142,6 +142,19 @@ export class VmFormComponent implements FormConfiguration {
           tooltip: helptext.threads.tooltip,
           validation: [Validators.required, Validators.min(1), this.cpuValidator('threads')],
         },
+        {
+          type: 'input',
+          name: 'cpuset',
+          inputType: 'string',
+          placeholder: helptext.cpuset.placeholder,
+          tooltip: helptext.cpuset.tooltip,
```
[the remark this answers is not in the document]
Comment (AlexKarpov98) [reply src/app/pages/vm/vm-form/vm-form.component.ts:149]: Also add validations for `cpuset`, `nodeset` fields in [vm-wizard.component.ts](https://github.com/truenas/webui/blob/c812ab457749c1073388ff9018698f759bc845c7/src/app/pages/vm/vm-wizard/vm-wizard.component.ts)

**Review by AlexKarpov98 — COMMENTED** [review]
LGTM!
Screenshot_78
Screenshot_79

File: src/app/pages/vm/vm-form/vm-form.component.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (heppu) [reply src/app/pages/vm/vm-form/vm-form.component.ts:149]: Where in the repo should I place this validator logic?

[the remark this answers is not in the document]
Comment (heppu) [reply src/app/pages/vm/vm-form/vm-form.component.ts:149]: Oh I did not notice your edit. I will just use inline regex then.

[the remark this answers is not in the document]
Comment (heppu) [reply src/app/pages/vm/vm-form/vm-form.component.ts:149]: Done.

**Comment by heppu** [comment]
Rebased to lates master and added validation. Should be now ready to merge.

**Review by AlexKarpov98 — CHANGES_REQUESTED** [review]
Seems validation does not work properly.
Screenshot_99
Screenshot_100

**Comment by heppu** [comment]
Hmm I don't have access to TrueNAS server right now to test against with so I used only online regex tool to validate it. I will update this once I get back home and can properly test it.

**Review by AlexKarpov98 — CHANGES_REQUESTED** [review]
This issue doesn't still work.

image

**Comment by heppu** [comment]
@AlexKarpov98 Should be now fixed. Didn't remember that \ had to be escaped in js :D

**Comment by heppu** [comment]
Ping @AlexKarpov98

**Review by undsoft — CHANGES_REQUESTED** [review]
New fields are not passed to websocket when form is saved.
You need to add:
```
    vmPayload['cpuset'] = value.cpuset;
    vmPayload['nodeset'] = value.nodeset;
    vmPayload['pin_vcpus'] = value.pin_vcpus;
```
in `VmWizardComponent` `customSubmit`.

**Comment by heppu** [comment]
@undsoft Done

**Review by undsoft — APPROVED** [review]
Looks good, thanks.

**Pull request merged**